include/deal.II/lac/precondition.h

```diff
@@ -2521,15 +2846,62 @@ PreconditionRelaxation<MatrixType, PreconditionerType>::Tstep(
   const VectorType &src) const
 {
   Assert(this->A != nullptr, ExcNotInitialized());
-  Assert(this->preconditioner != nullptr, ExcNotInitialized());
+  Assert(data.preconditioner != nullptr, ExcNotInitialized());
+
+  if (eigenvalues_are_initialized == false)
+    estimate_eigenvalues(src);
 
   VectorType tmp1, tmp2;
 
-  for (unsigned int i = 1; i <= n_iterations; ++i)
+  for (unsigned int i = 1; i <= data.n_iterations; ++i)
     internal::PreconditionRelaxation::step_operations(
-      *A, *preconditioner, dst, src, relaxation, tmp1, tmp2, i, true);
+      *A, *data.preconditioner, dst, src, data.relaxation, tmp1, tmp2, i, true);
 }
 
+template <typename MatrixType, typename PreconditionerType>
+template <typename VectorType>
+inline internal::EigenvalueInformation
+PreconditionRelaxation<MatrixType, PreconditionerType>::estimate_eigenvalues(
+  const VectorType &src) const
+{
+  Assert(eigenvalues_are_initialized == false, ExcInternalError());
+
```

Comment:
Also here it appears as if we have redundancy and copy-paste from the other code. Could you try to unify more?

Reply:
The relevant lines of codes are already shared via the internal function `internal::estimate_eigenvalues`. `PreconditionRelaxation` and `PreconditionChebyshev` need to do some additional/different steps since 1) `PreconditionRelaxation` does not store the vectors and 2) the estimated eigenvalues are used differently. What would you move additionally to the internal function?